snuba/query/query.py

```diff
@@ -35,6 +40,23 @@
 TElement = TypeVar("TElement")
 
 
+class OrderByDirection(Enum):
+    ASC = "asc"
+    DESC = "desc"
+
+
+@dataclass(frozen=True)
+class OrderBy:
+    direction: OrderByDirection
+    node: Expression
+
+    def replace_node(self, new_node: Expression) -> OrderBy:
+        """
+        Returns a new OrderBy clause with a new node.
+        """
+        return OrderBy(self.direction, new_node)
```

Comment:
Can callers use [`dataclass.replace`](https://docs.python.org/3/library/dataclasses.html#dataclasses.replace) instead of this function, or does that have typing implications? Looking at the signature of `replace` I could see how some type information would be lost (but I haven't tested this.)

I general I think actually exposing a `replace` method here (and on `Expression`) that binds the instance argument of `replace` to `self` would be the nicest API from a caller's perspective. I suspect that this pattern will be used a lot in `transform` functions.

Reply:
For now I relied on dataclasses.replace because, if I defined the replace method in the classes themselves, I did not manage to convince mypy that the types made sense.